### PR TITLE
Resolve presence race and introduce new Presence API

### LINKED ISSUE #2255: Race condition when presence is left shortly after presence state retrieval
### Environment

* Elixir version (elixir -v): Elixir 1.4.0, erts-8.2)
* Phoenix version (mix deps): phoenix 1.3.0-rc.1
* NodeJS version (node -v): v7.7.2
* NPM version (npm -v): 4.4.1
* Operating system: OS X 10.12.3

### Expected behavior

Using the default setup for channels and presence, any leaves happening concurrently with a new channel's connection should result in a client side presence state without the left presence (either through it not being included in `presence_state` or it being removed shortly thereafter through a `presence_diff`.

### Actual behavior

Some presence changes, that are applied after the newly joined channel has retrieved its current presence state, will be sent to the client before the presence_state event is sent. When this happens, any such presence leaves will be ignored by the default js syncDiff. Once the presence_state event is received, the left presences will be kept even though they are not present anymore.

### Easiest way to reproduce locally

1. Introduce a delay in a `Presence.fetch/2` callback, that scales linearly with the number of presences. Example: `def fetch(_topic, presences) do Process.sleep(1000 * Enum.count presences); presences end`
2. Open a few tabs to make sure the presence state contains 5 or more items.
3. Reload a tab, triggering a new channel connection.
4. Within a few seconds, close one or more of the other tabs.
5. Observe that the newly opened tab recevies the leave events before it's presence_state, and that it keeps those left presences until the next reload.

A runnable example of the above is available at https://github.com/alvinlindstam/phoenix_social_presence/tree/race-condition

I understand that the one second delay is an extreme example, but a fetch function that takes longer for many presences than for a small diff is hardly unlikely. I'm not sure if this would be possible to reproduce without a fetch function, but I can't se any reason why a presence change handled just after a presence state fetch by a new channel couldn't be sent to and handled by the transport process before the channel process itself sends the state to the transport.

### PR DESCRIPTION
**Pull request by chrismccord**
Fixes #2255, for real this time :)

@josevalim and @alvinlindstam I would appreciate thoughts on this new API. It keeps the "lower level" existing API and simplifies the knowledge and code required to handle presence events. At the same time it also resolves the race condition in #2255 by tracking the joinRef or "session id" of the channel connection. The new API looks like this (taken from Alvin's example app):

```javascript
const renderPresences = (presence) => {
  const listed_presences = presence.list((user, {metas: metas}) => {
    return {user: user, ...}
  })
  ...
}

let channel = socket.channel("user:all", {})
let presence = new Presence(channel)

presence.onSync(() => {
  renderPresences(presence)
})

channel.join()...
```

Note: I still need to update the presence.js docs with the new API, but I'm awaiting feedback. Thanks!

### CODE REVIEW

**Comment by OvermindDL1** [comment]
What about when you have different presences coming down a single channel (as I do)?

**Comment by chrismccord** [comment]
Not sure I follow? How are you handling it today that this api would not allow? (Keep in mind the "old" api is still there, but you would need to resolve the race yourself)

**Comment by OvermindDL1** [comment]
The standard naming conventions that presence uses by default I don't at all, thus if I wanted to use the new API defined here I'd also imagine being able to pass in different names other than `presence_state` and `presence_diff` as, for example, I instead use `notifications_state`/`notifications_diff` and `members_state`/`members_diff`.  I.E. having the ability to pass in different message names than the ones it is currently hard-coded to would let me use this new API (and remove all my manual synching code, whoo).  :-)

Basically I want to be able to do this:

```javascript
let notifications_presence = new Presence(channel, "notifications_state", "notifications_diff")
```

Or maybe just this, which would just alias to the above (for the simple case where you follow the `<name>_state`/`<name>_diff` pattern as I do):

```javascript
let notifications_presence = new Presence(channel, "notifications")
```

**Comment by chrismccord** [comment]
Oh I see. We can easily support those as options. Thanks!

**Comment by jeregrine** [comment]
👍 I don't love that it uses internal state to keep track of it. But this is JavaScript, and its probably best to keep with how they handle it.

**Comment by josevalim** [comment]
@chrismccord do we have docs for update? Also, do we generate some javascript in the generators or do we show some javascript in them? If so they need to be updated too.

**Comment by chrismccord** [comment]
@josevalim I was waiting on feedback for the API before finalizing docs. If all looks good from your side I'll touch up the docs

**Comment by josevalim** [comment]
Good for me!

**Comment by OvermindDL1** [comment]
@chrismccord And just a reminder ping on the message naming options.  ^.^

**Comment by christianevans214** [comment]
Hey there! Sorry to butt in on this PR, but is there an ETA on when this new presence API will be out via NPM?
